server/aws-lsp-antlr4/src/language-service/service.ts

```diff
@@ -0,0 +1,150 @@
+import {
+    Diagnostic,
+    Hover,
+    DiagnosticSeverity,
+    TextDocument,
+    CompletionList,
+    CompletionItem,
+} from '@aws/language-server-runtimes/server-interface'
+
+import { type AwsLanguageService } from '@aws/lsp-core'
+import {
+    CommonTokenStream,
+    Token,
+    CharStream,
+    ATNSimulator,
+    Recognizer,
+    BaseErrorListener,
+    type Lexer,
+    type Parser,
+} from 'antlr4ng'
+import { getPosition, getTokenIndexFromParseTree, getCandidates } from './utils/completion'
+import type { TextEdit } from 'vscode-languageserver-textdocument'
+
+export function normalizeQuery(data: string): string {
+    return data != null ? data : ''
+}
+
+export function createANTLR4LanguageService(
+    supportedLanguageIds: string[],
+    antlrLexerConstructor: (charStream: CharStream) => Lexer,
+    antlrParserConstructor: (tokenStream: CommonTokenStream) => Parser,
+    parserRoot: string,
+    ignoredTokens?: number[]
+) {
+    return new ANTLR4LanguageService(
+        supportedLanguageIds,
+        antlrLexerConstructor,
+        antlrParserConstructor,
+        parserRoot,
+        ignoredTokens
+    )
+}
+
+// Collector used to give to the ANTLR parser and lexer to store the diagnostics.
+class DiagnosticsCollector extends BaseErrorListener {
+    private _diagnostics: Diagnostic[] = []
+
+    constructor(private _document: TextDocument) {
+        super()
+    }
+
+    // We only use the parser for diagnostics, so we're only interested in the syntaxError method for this class.
+    syntaxError(recognizer: Recognizer<ATNSimulator>, offendingSymbol: Token | null) {
+        this._diagnostics.push({
+            message: `Unexpected token: ${offendingSymbol?.text}`,
+            range: {
+                start: this._document.positionAt(offendingSymbol?.start ?? 0),
+                end: this._document.positionAt(offendingSymbol?.stop ?? 0),
+            },
+            severity: DiagnosticSeverity.Error,
+        })
+    }
+
+    get errors() {
+        return this._diagnostics
+    }
+}
+
+export class ANTLR4LanguageService implements AwsLanguageService {
+    private lexerConstructor
+    private parserConstructor
+    protected parserRoot
+    protected ignoredTokens
+
+    constructor(
+        private supportedLanguageIds: string[],
+        lexerConstructor: (charStream: CharStream) => Lexer,
+        parserConstructor: (tokenStream: CommonTokenStream) => Parser,
+        parserRoot: string,
+        ignoredTokens?: number[]
+    ) {
+        this.lexerConstructor = lexerConstructor
+        this.parserConstructor = parserConstructor
+        this.parserRoot = parserRoot
+        this.ignoredTokens = new Set(ignoredTokens ?? [])
+    }
+    isSupported(document: TextDocument): boolean {
```

Comment:
Do we use this anywhere?

Reply:
this is implemented via the interface and used internally